use BOOL for TCP_NODELAY setsockopt value on Windows

chrisnc opened the pull request:
This issue was found by the Wine project and mitigated there [^1].

Windows' setsockopt expects a BOOL (a typedef for int) for TCP_NODELAY
[^2]. Windows itself is forgiving and will accept any positive optlen and
interpret the first byte of *optval as the value, so this bug does not
affect Windows itself, but does affect systems implementing Windows'
interface more strictly, such as Wine. Wine was previously passing this
through to the host's setsockopt, where, e.g., Linux requires that
optlen be correct for the chosen option, and TCP_NODELAY expects an int.

[^1]: https://source.winehq.org/git/wine.git/commit/d6ea38f32dfd3edbe107a255c37e9f7f3da06ae7
[^2]: https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-setsockopt

chrisnc committed:
use `BOOL` for `TCP_NODELAY` `setsockopt` value on Windows

This issue was found by the Wine project and mitigated there [1].

Windows' documented interface for `setsockopt` expects a `BOOL` (a
`typedef` for `int`) for `TCP_NODELAY` [2]. Windows is forgiving and
will accept any positive length and interpret the first byte of
`*option_value` as the value, so this bug does not affect Windows
itself, but does affect systems implementing Windows' interface more
strictly, such as Wine. Wine was previously passing this through to the
host's `setsockopt`, where, e.g., Linux requires that `option_len` be
correct for the chosen option, and `TCP_NODELAY` expects an `int`.

[1]: https://source.winehq.org/git/wine.git/commit/d6ea38f32dfd3edbe107a255c37e9f7f3da06ae7
[2]: https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-setsockopt

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @dtolnay (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

chrisnc commented:
Fixed footnote index...

dtolnay reviewed (approved):
Thanks!

chrisnc commented:
I guess I didn't properly link this commit to #94092, since I made the commit before creating the issue. Should I update the commit message again to do so, or can someone with write perms link it for me? https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#manually-linking-a-pull-request-to-an-issue

matthiaskrgr commented:
@bors r-
looks like this causes problems on ci:  https://github.com/rust-lang/rust/pull/94152#issuecomment-1046048346

chrisnc commented:
Agh, looks like the assertion in the getsockopt wrapper is complaining about the size. The docs say that the optval for TCP_NODELAY on getsockopt should also be BOOL (int), but this test failure suggests that getsockopt is setting *optlen to 1... I will investigate.

chrisnc committed:
Fix `setsockopt` and `getsockopt` parameter names.

Previously `level` was named `opt` and `option_name` was named `val`,
then extra names of `payload` or `slot` were used for the option value.
This change aligns the wrapper parameters with their names in POSIX.
Winsock uses similar but more abbreviated names: `level`, `optname`,
`optval`, `optlen`.

chrisnc committed:
Remove assertion on output length for `getsockopt`.

POSIX allows `getsockopt` to set `*option_len` to a smaller value if
necessary. Windows will set `*option_len` to 1 for boolean options even
when the caller passes a `BOOL` (`int`) with `*option_len` as 4.

chrisnc commented:
So, I had tested the getsockopt change previously, but I was only testing on Wine, which does not modify `*optlen` and hence doesn't trigger this `assert_eq`... This makes me think that this assertion is too strict. POSIX allows for `*optlen` to be set to a lower value than what is passed in, and Windows is clearly doing this in reality (setting it to 1 after 4 is passed in). It worked before because Rust was always passing in 1, which strictly speaking is not correct but Windows accepts anyway. What do you think @dtolnay?

chrisnc commented:
Somewhat related, I noticed that the `getsockopt` and `setsockopt` wrappers in `sys_common/net.rs` have incorrect argument names, compared with the corresponding names on the C interface. What the wrapper calls `opt` is actually `level`, `val` is actually `option_name`, and then because it already used `val` for the wrong thing, it calls the option value `payload` in `setsockopt` and `slot` in `getsockopt`. I'd like to change these.

ChrisDenton commented:
So to sum up the issue:

* The data for `TCP_NODELAY` is documented as being a `BOOL` (a 32 bit signed integer)
* The Windows implementation will only actually read or write the first byte.
* WINE used the full 4 bytes.
* The specification for `getsockopt`/`setsockopt` is flexible and allows for either implementation.

Is that about right?

chrisnc commented:
@ChrisDenton yes that's most of it. Prior to the recent patch, Wine was expecting 4 bytes for the `setsockopt`. For `getsockopt`, I believe it was also expecting 4 bytes but leaving *optlen unchanged.
Windows will accept >= 1 byte for `setsockopt` and `getsockopt`, and then for `getsockopt`, it will set `*optlen` to 1, even if it was passed in as 4. (This is evident from the failing assertion in the test; I'm still working on getting my Windows environment set up and a working example to reproduce this.) I've been in contact with the Wine maintainers who made the recent change, and they're also interested in matching the behavior of Windows on the final value of `*optlen` for `getsockopt`. I have patches that fix up the parameter names and one to remove the assertion, but I wanted to get some feedback on whether this is the right path. I can push them here for review shortly.

chrisnc commented:
I've put two commits ahead of the original one. Appreciate any feedback, particularly on removing the assert.

chrisnc commented:
r? rust-lang/libs

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit b02698c7e6843d6feacc394cb7f83f3fc347c3e2 has been approved by `dtolnay`

<!-- @bors r=dtolnay b02698c7e6843d6feacc394cb7f83f3fc347c3e2 -->
<!-- homu: {"type":"Approved","sha":"b02698c7e6843d6feacc394cb7f83f3fc347c3e2","approver":"dtolnay"} -->
